### ISSUE TITLE
Modules are not properly loaded with Jest

### ISSUE THREAD

**Issue by rubensworks** [RubenVerborgh/N3.js#120]
In certain environments (such as Jest's test environment), N3 isn't loaded properly.

This seems to be caused by N3's custom lazy require loader. In Jest's environment, all modules (such as `N3Util`) seem to be an empty hash.
I suspect that Jest is hooking into the setter of `module.exports`, which causes N3's custom lazy-loading behaviour to be ignored.

Removing this lazy loading logic resolves the problem, but I suspect this is not something you want?

**Comment by RubenVerborgh** [comment]
Would there be another lazy loading mechanism that is compatible with Jest?

Or maybe the lazy loading doesn't make a big difference anymore; it used to for some benchmarks (for instance, loading parser when only store is needed).

**Comment by rubensworks** [comment]
I'll have a look tomorrow to see if I can somehow make it work with Jest. It could possibly be done by simply moving the assignment of `module.exports` to the end.

**Comment by rubensworks** [comment]
I've tried several things, but no other lazy loading mechanisms seem to work.

It looks like the issue is know (facebook/jest#872, facebook/jest#2605), but apparently not something they consider an issue on their end, which it obviously is.

I think they may be somehow disabling `require` while the tests are running, because a change like this apparently resolves it: https://github.com/sidorares/node-mysql2/issues/489#issuecomment-313374683

TL;DR: An issue with Jest, but they don't acknowledge it.

**Comment by RubenVerborgh** [comment]
Maybe it's time to drop the lazy loading, I'm not sure. Feel free to try!

**Comment by rubensworks** [comment]
I'm in favor of lazy loading to be honest, so I wouldn't drop this functionality just because Jest doesn't want to fix their issues :-)

I'll be needing lazy require's in any case in the context of Jest testing in any case, so a fix on their end is needed.

**Comment by rubensworks** [comment]
For reference, this can be fixed by including the following at the top of your Jest test file:

```
const n3 = require('n3');
Object.keys(n3).forEach((submodule) => n3[submodule] = n3[submodule]);
```

**Comment by RubenVerborgh** [comment]
Ooh, nice.

If you're happy, we can close this one.

**Comment by RubenVerborgh** [comment]
(Actually, I suspect `for (var key in n3) n3[key];` would also work.)

**Comment by rubensworks** [comment]
Indeed, `for (var key in n3) n3[key];` also works, Jest doesn't seem to be breaking that one either :-)

**Comment by RubenVerborgh** [comment]
For future reference: removed lazy loading in ffba1ae52c9a3fff2065c15cdac62a2c622ab7e4
